stan/math/prim/mat/prob/poisson_log_glm_lpmf.hpp

```diff
@@ -0,0 +1,143 @@
+     * length of n and whose number of columns should match the length of beta
+     * @tparam T_beta type of the weight vector;
+     * this can also be a single value;
+     * @tparam T_alpha type of the intercept;
+     * this should be a single value;
+     * @param n positive integer vector parameter
+     * @param x design matrix
+     * @param beta weight vector
+     * @param alpha intercept (in log odds)
+     * @return log probability or log sum of probabilities
+     * @throw std::domain_error if x, beta or alpha is infinite.
+     * @throw std::domain_error if n is negative.
+     * @throw std::invalid_argument if container sizes mismatch.
+     */
+    template <bool propto, typename T_n, typename T_x, typename T_beta,
+              typename T_alpha>
+    typename return_type<T_x, T_beta, T_alpha>::type
+    poisson_log_glm_lpmf(const T_n &n, const T_x &x, const T_beta &beta,
+                             const T_alpha &alpha) {
+      static const std::string function = "poisson_log_glm_lpmf";
+      typedef typename stan::partials_return_type<T_n, T_x, T_beta,
+                                                  T_alpha>::type
+        T_partials_return;
+
+      using std::exp;
+      using Eigen::Dynamic;
+      using Eigen::Matrix;
+
+      if (!(stan::length(n) && stan::length(x) && stan::length(beta)))
+        return 0.0;
+
+      T_partials_return logp(0.0);
+
+      check_nonnegative(function, "Vector of dependent variables", n);
+      check_finite(function, "Matrix of independent variables", x);
+      check_finite(function, "Weight vector", beta);
+      check_finite(function, "Intercept", alpha);
+      check_consistent_sizes(function,
+                             "Rows in matrix of independent variables",
+                             x.col(0), "Vector of dependent variables",  n);
+      check_consistent_sizes(function,
+                             "Columns in matrix of independent variables",
+                             x.row(0), "Weight vector",  beta);
+
+      if (!include_summand<propto, T_x, T_beta, T_alpha>::value)
+        return 0.0;
+
+      const size_t N = x.col(0).size();
+      const size_t M = x.row(0).size();
+
+      Matrix<T_partials_return, Dynamic, 1> n_vec(N, 1);
+      {
+        scalar_seq_view<T_n> n_seq_view(n);
+        for (size_t n = 0; n < N; ++n) {
+          n_vec[n] = n_seq_view[n];
+        }
+      }
+
+      Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
+      {
+        scalar_seq_view<T_beta> beta_vec(beta);
+        for (size_t m = 0; m < M; ++m) {
+          beta_dbl[m] = value_of(beta_vec[m]);
+        }
+      }
+      Matrix<T_partials_return, Dynamic, Dynamic> x_dbl = value_of(x);
+
+      Matrix<T_partials_return, Dynamic, 1> theta_dbl = (x_dbl * beta_dbl
+        + Matrix<double, Dynamic, 1>::Ones(N, 1)
+        * value_of(alpha));
+      Matrix<T_partials_return, Dynamic, 1> exp_theta =
+        theta_dbl.array().exp().matrix();
```

Comment:
Is conversion to array and back to matrix the best way to do this? I'm not sure of the performance implications -does it copy or make a new view?

Reply:
I think so. There are no runtime implications. See https://eigen.tuxfamily.org/dox/group__TutorialArrayClass.html . "Matrix expressions have an .array() method that 'converts' them into array expressions, so that coefficient-wise operations can be applied easily. Conversely, array expressions have a .matrix() method. As with all Eigen expression abstractions, this doesn't have any runtime cost (provided that you let your compiler optimize)."

Reply:
Very cool, thanks!